include/maths/CBoostedTreeFactory.h

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+#ifndef INCLUDED_ml_maths_CBoostedTreeFactory_h
+#define INCLUDED_ml_maths_CBoostedTreeFactory_h
+
+#include <core/CContainerPrinter.h>
+#include <core/CDataFrame.h>
+#include <core/CPackedBitVector.h>
+
+#include <maths/CBayesianOptimisation.h>
+#include <maths/CBoostedTree.h>
+
+#include <memory>
+#include <numeric>
+#include <sstream>
+#include <utility>
+#include <vector>
+
+namespace ml {
+namespace maths {
+
+class CNode;
+class CBoostedTree;
+
+//! SimpleFactory for CBoostedTree object
+class CBoostedTreeFactory final {
+public:
+    using TBoostedTreeUPtr = std::unique_ptr<CBoostedTree>;
+
+public:
+    //! Construct a boosted tree object from parameters
+    static CBoostedTreeFactory constructFromParameters(std::size_t numberThreads,
+                                                       std::size_t dependentVariable,
+                                                       CBoostedTree::TLossFunctionUPtr loss);
+
+    //! Set the number of folds to use for estimating the generalisation error.
+    CBoostedTreeFactory& numberFolds(std::size_t folds);
+    //! Set the lambda regularisation parameter.
+    CBoostedTreeFactory& lambda(double lambda);
+    //! Set the gamma regularisation parameter.
+    CBoostedTreeFactory& gamma(double gamma);
+    //! Set the amount we'll shrink the weights on each each iteration.
+    CBoostedTreeFactory& eta(double eta);
+    //! Set the maximum number of trees in the ensemble.
+    CBoostedTreeFactory& maximumNumberTrees(std::size_t maximumNumberTrees);
+    //! Set the fraction of features we'll use in the bag to build a tree.
+    CBoostedTreeFactory& featureBagFraction(double featureBagFraction);
+    //! Set the maximum number of optimisation rounds we'll use for hyperparameter
+    //! optimisation per parameter.
+    CBoostedTreeFactory& maximumOptimisationRoundsPerHyperparameter(std::size_t rounds);
+    //! Set the callback function for progress monitoring.
+    CBoostedTreeFactory& progressCallback(CBoostedTree::TProgressCallback callback);
+    //! Set the reference to the data frame.
+    CBoostedTreeFactory& frame(core::CDataFrame& frame);
```

Comment:
This underlines why I'm not sold on having memory estimation mixed up with having to build an object: this is a non-optional argument and so code will have to check if it has been specified and then you only get runtime rather than compile time failures for misuse.

I'd much prefer parameters which are not optional, such as this one, to be passed in the `constructFromParameters` method for this reason.

Reply:
`frame` is  now a required parameter for the constructor of `CBoostedTree`. Unfortunately, it cannot be passed over `constructFromParameters` since `CBoostedTreeRunner` does not have it at the moment when `constructFromParameters` is invoked. I introduced a check int the `CBoostedTreeFactory::build()` to prevent constructing a `CBoostedTree` object without `frame` initialized.